sidekick_core/lib/src/sidekick_context.dart

```diff
@@ -0,0 +1,316 @@
+      final startDir = knownSidekickPackage?.root ?? script.parent;
+
+      SidekickPackage? sidekickPackage = knownSidekickPackage;
+      EntryPoint? entryPoint;
+      ProjectDiscoveryResult? result;
+      for (final dir in startDir.allParentDirectories()) {
+        sidekickPackage ??= SidekickPackage.fromDirectory(dir);
+        entryPoint ??= () {
+          if (sidekickPackage == null) {
+            // sidekickPackage is always found first, when searching upwards
+            return null;
+          }
+          final String entryPointName = sidekickPackage.cliName;
+          final match = dir
+              .listSync()
+              .whereType<File>()
+              .firstOrNullWhere((it) => it.name == entryPointName);
+          if (match != null) {
+            return EntryPoint(file: match);
+          }
+          return null;
+        }();
+        if (sidekickPackage != null && entryPoint != null) {
+          result = ProjectDiscoveryResult(
+            sidekickPackage: sidekickPackage,
+            entryPoint: entryPoint,
+          );
+          break;
+        }
+      }
+      if (result == null) {
+        throw StateError(
+          "Can't find sidekick package and entryPoint from ${script.path}.\n"
+          'entryPoint: $entryPoint,\n'
+          'sidekickPackage: $sidekickPackage',
+        );
+      }
+      return result;
+    });
+  }
+
+  /// The directory where the [entryPoint] is located
+  ///
+  /// This directory is considered the project root.
+  static Directory get projectRoot {
+    if (env.exists(_envEntryPointHome)) {
+      // CLI is called via entryPoint
+      final injectedEntryPointPath = env[_envEntryPointHome];
+      if (injectedEntryPointPath == null || injectedEntryPointPath.isBlank) {
+        throw 'Injected entryPoint was not set (env.$_envEntryPointHome)';
+      }
+      final Directory dir = Directory(injectedEntryPointPath);
+      if (!dir.existsSync()) {
+        throw 'Injected entryPoint does not exist ${dir.absolute.path}';
+      }
+      return dir;
+    }
+    return entryPoint.file.parent;
+  }
+
+  /// The git repository root the [sidekickPackage] is located in
+  @Deprecated('Use projectRoot instead')
+  static Repository get repository {
+    // TODO remove repository? It's not actually required when we define
+    //  the entryPoint to be the root of the project and start package discovery
+    //  from there.
+    return _cache.getOrCreate('findRepository', _findRepository);
+  }
+
+  static Repository _findRepository() {
+    bool isGitDir(Directory dir) => dir.directory('.git').existsSync();
+
+    final projectRoot = entryPoint.file.parent;
+    final gitRoot = projectRoot.findParent(isGitDir);
+    if (gitRoot == null) {
+      throw 'Could not find the root of the repository from ${projectRoot.path}';
+    }
+    return Repository(root: gitRoot);
+  }
+}
+
+SidekickContextCache get internalSidekickContextCache {
```

Comment:
Users probably shouldn't access this, right? If so we could add a `@protected` annotation to this getter and the corresponding setter

Reply:
it's not exported, therefore unlikely someone will use it